core/node/base_token_adjuster/src/base_token_ratio_persister.rs

```diff
@@ -1,72 +1,82 @@
-use std::{fmt::Debug, num::NonZero};
+use std::{fmt::Debug, sync::Arc};
 
 use anyhow::Context as _;
-use chrono::Utc;
 use tokio::sync::watch;
 use zksync_config::configs::base_token_adjuster::BaseTokenAdjusterConfig;
 use zksync_dal::{ConnectionPool, Core, CoreDal};
-use zksync_types::base_token_ratio::BaseTokenAPIRatio;
+use zksync_external_price_api::PriceAPIClient;
+use zksync_types::{base_token_ratio::BaseTokenAPIRatio, Address};
 
 #[derive(Debug, Clone)]
 pub struct BaseTokenRatioPersister {
     pool: ConnectionPool<Core>,
     config: BaseTokenAdjusterConfig,
+    base_token_address: Address,
+    price_api_client: Arc<dyn PriceAPIClient>,
 }
 
 impl BaseTokenRatioPersister {
-    pub fn new(pool: ConnectionPool<Core>, config: BaseTokenAdjusterConfig) -> Self {
-        Self { pool, config }
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        config: BaseTokenAdjusterConfig,
+        base_token_address: Address,
+        price_api_client: Arc<dyn PriceAPIClient>,
+    ) -> Self {
+        Self {
+            pool,
+            config,
+            base_token_address,
+            price_api_client,
+        }
     }
 
     /// Main loop for the base token ratio persister.
     /// Orchestrates fetching a new ratio, persisting it, and conditionally updating the L1 with it.
     pub async fn run(&mut self, mut stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
         let mut timer = tokio::time::interval(self.config.price_polling_interval());
-        let pool = self.pool.clone();
 
         while !*stop_receiver.borrow_and_update() {
             tokio::select! {
                 _ = timer.tick() => { /* continue iterations */ }
                 _ = stop_receiver.changed() => break,
             }
 
-            let new_ratio = self.fetch_new_ratio().await?;
-            self.persist_ratio(&new_ratio, &pool).await?;
-            // TODO(PE-128): Update L1 ratio
+            if let Err(err) = self.loop_iteration().await {
+                tracing::error!(
+                    "Failed to execute a base_token_ratio_persister loop iteration: {err}"
+                );
+            }
```

Comment:
Do we consider any error to be transient? Normally we treat database errors as irrecoverable and bail, while trying to recover on transient errors only. Otherwise there may be potentially dangerous situations; e.g. imagine that tx for persisting token on L1 reverts -- we will be sending a reverting transaction over and over, draining the operator's balance.

I'm raising it now, because later these lines won't be a part of diff, and it'll be harder to notice the problem.

Reply:
Good point. Made it a persistent error and added a retry mechanism in order to not shut down & alert whenever Coingecko temporarily goes down.